### PR TITLE
refactored test-end-to-end/core to use os::cmd functions

### PR DESCRIPTION
**Pull request by stevekuznetsov**
@liggitt PTAL

### CODE REVIEW

File: test/end-to-end/core.sh [hunk 1]
```diff
@@ -18,55 +19,55 @@ TEST_ASSETS="${TEST_ASSETS:-false}"
 function wait_for_app() {
   echo "[INFO] Waiting for app in namespace $1"
   echo "[INFO] Waiting for database pod to start"
-  wait_for_command "oc get -n $1 pods -l name=database | grep -i Running" $((60*TIME_SEC))
-  oc logs dc/database -n $1 --follow
+  os::cmd::try_until_text "oc get -n $1 pods -l name=database" 'Running'
+  os::cmd::expect_success 'oc logs dc/database -n $1 --follow' # TODO(skuznets): determine why we use --follow here
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @liggitt Do we want to grab the build logs at a later point or do we need `--follow` here for some reason?

File: test/end-to-end/core.sh [hunk 2]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
...
-oc whoami
+os::cmd::expect_success 'oc login -u e2e-user'
+os::cmd::expect_success 'oc project test'
+os::cmd::expect_success 'oc whoami'
 
 echo "[INFO] Running a CLI command in a container using the service account"
-oc policy add-role-to-user view -z default
+os::cmd::expect_success 'oc policy add-role-to-user view -z default'
 out=$(oc run cli-with-token --attach --env=POD_NAMESPACE=test --image=openshift/origin:${TAG} --restart=Never -- cli status --loglevel=4 2>&1)
-echo "${out}"
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @liggitt Since we run everything in a subshell, I don't see how we can get the pesky `out=$()` lines into a wrapper function. Is this something we can run a couple of times? I guess this is a second use-case for testing a command's output for multiple strings.

File: test/end-to-end/core.sh [hunk 3]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
...
-os::build:wait_for_start "test"
+os::cmd::wait_for_text 'oc get -n test builds' 'Running'
+ 
 # Ensure that the build pod doesn't allow exec
-[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
-os::build:wait_for_end "test"
+BUILD_ID=$(oc get -n $1 builds  --output-version=v1 --template="{{with index .items 0}}{{.metadata.name}}{{end}}")
+os::cmd::expect_failure_and_text "oc rsh ${BUILD_ID}-build" 'forbidden'
+os::cmd::try_until_text 'oc get -n text builds' 'Complete'
+os::cmd::expect_success "oc build-logs -n test ${BUILD_ID} > ${LOG_DIR}/testbuild.log" # TODO(skuznets): previously we were agnostic about this failing, why?
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: In `os::build::wait_for_end` we were OK with this failing. Why?

File: test/end-to-end/core.sh [hunk 4]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
...
-[ "$(oc rsh ${frontend_pod} id -u | grep 1000)" ]
-[ "$(oc rsh -T ${frontend_pod} id -u | grep 1000)" ]
+os::cmd::expect_success_and_text "oc exec -p ${frontend_pod} id" '1000'
+os::cmd::expect_success_and_text "oc rsh ${frontend_pod} id -u" '1000'
+os::cmd::expect_success_and_text "oc rsh -T ${frontend_pod} id -u" '1000'
 
 # Port forwarding
 echo "[INFO] Validating port-forward"
-oc port-forward -p ${frontend_pod} 10080:8080  &> "${LOG_DIR}/port-forward.log" &
+os::cmd::expect_success "oc port-forward -p ${frontend_pod} 10080:8080  &> ${LOG_DIR}/port-forward.log &" # TODO(skuznets): why end this with &?
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Unclear if this background run will play nice with the wrapper function.

File: test/end-to-end/core.sh [hunk 5]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
...
-[ "$(oc rsh ${frontend_pod} id -u | grep 1000)" ]
-[ "$(oc rsh -T ${frontend_pod} id -u | grep 1000)" ]
+os::cmd::expect_success_and_text "oc exec -p ${frontend_pod} id" '1000'
+os::cmd::expect_success_and_text "oc rsh ${frontend_pod} id -u" '1000'
+os::cmd::expect_success_and_text "oc rsh -T ${frontend_pod} id -u" '1000'
 
 # Port forwarding
 echo "[INFO] Validating port-forward"
-oc port-forward -p ${frontend_pod} 10080:8080  &> "${LOG_DIR}/port-forward.log" &
+os::cmd::expect_success "oc port-forward -p ${frontend_pod} 10080:8080  &> ${LOG_DIR}/port-forward.log &"
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Unclear how the background run will play into this.

File: test/end-to-end/core.sh [hunk 6]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
 # services can end up on any IP.  Make sure we get the IP we need for the docker registry
 DOCKER_REGISTRY=$(oc get --output-version=v1beta3 --template="{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}" service docker-registry)
 
-registry="$(dig @${API_HOST} "docker-registry.default.svc.cluster.local." +short A | head -n 1)"
-[[ -n "${registry}" && "${registry}:5000" == "${DOCKER_REGISTRY}" ]]
+os::cmd::expect_success_and_text "dig @${API_HOST} docker-registry.default.svc.cluster.local. +short A | head -n 1 | xargs printf '%s:5000'" "${DOCKER_REGISTRY}"
 
 echo "[INFO] Verifying the docker-registry is up at ${DOCKER_REGISTRY}"
-wait_for_url_timed "http://${DOCKER_REGISTRY}/healthz" "[INFO] Docker registry says: " $((2*TIME_MIN))
+os::cmd::try_until_success "curl --max-time 2 -fs http://${DOCKER_REGISTRY}/healthz" $((2*TIME_MIN))
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: I fail locally here every time.

File: test/end-to-end/core.sh (hunk 2 above)
Comment (liggitt) [inline test/end-to-end/core.sh]: definitely shouldn't run it multiple times (it's expensive)

File: test/end-to-end/core.sh [hunk 7]
```diff
@@ -84,191 +85,192 @@ wait_for_registry
...
-oc project test
-oc whoami
+os::cmd::expect_success 'oc login -u e2e-user'
+os::cmd::expect_success 'oc project test'
+os::cmd::expect_success 'oc whoami'
 
 echo "[INFO] Running a CLI command in a container using the service account"
-oc policy add-role-to-user view -z default
+os::cmd::expect_success 'oc policy add-role-to-user view -z default'
 out=$(oc run cli-with-token --attach --env=POD_NAMESPACE=test --image=openshift/origin:${TAG} --restart=Never -- cli status --loglevel=4 2>&1)
```
Comment (liggitt) [inline test/end-to-end/core.sh]: seeing this output is the main reason I would care about refactoring this... probably won't get to reviewing the rest until after break

File: test/end-to-end/core.sh (hunk 1 above)
Comment (liggitt) [inline test/end-to-end/core.sh]: `--follow` streams logs until the deployment completes, which we want

File: test/end-to-end/core.sh (hunk 5 above)
Comment (liggitt) [inline test/end-to-end/core.sh]: don't wrap it, then... we're capturing the log in case it fails.

File: test/end-to-end/core.sh (hunk 7 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Can we tolerate running the command twice to test the output twice within a wrapper? We can't set the output to a variable in  a wrapper, we run in a subshell.

Comment (liggitt) [inline test/end-to-end/core.sh]: let's start with a separate PR that directs this output to `${LOG_DIR}/run-cli-with-token.log`, then cats that into a var for comparisons... I just want to see failure details

Comment (liggitt) [inline test/end-to-end/core.sh]: same for other `oc run` commands that capture output for comparisons

File: test/end-to-end/core.sh [hunk 8]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
-os::build:wait_for_start "test"
+os::cmd::try_until_text 'oc get -n test builds' 'Running'
+ 
 # Ensure that the build pod doesn't allow exec
-[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
-os::build:wait_for_end "test"
+BUILD_ID=$(oc get -n test builds  --output-version=v1 --template="{{with index .items 0}}{{.metadata.name}}{{end}}")
+os::cmd::expect_failure_and_text "oc rsh ${BUILD_ID}-build" 'forbidden'
+os::cmd::try_until_text 'oc get -n test builds' 'Complete' $((10*TIME_MIN))
+os::cmd::expect_success "oc build-logs -n test ${BUILD_ID} > ${LOG_DIR}/testbuild.log" # TODO(skuznets): previously we were agnostic about this failing, why?
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Previous comment got lost. In the previous impl of this, this log call was inside a `set +e` block. Why were we agnostic to it failing?

File: test/end-to-end/core.sh (hunk 7 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: That pull created here: https://github.com/openshift/origin/pull/6483

File: test/end-to-end/core.sh (hunk 8 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @deads2k this was you addition [here](https://github.com/openshift/origin/commit/e781d5e8f0a6cf44a2c80f307b9f381828c18217#diff-b83ac5455d252fc0916d54924166ffa5R523). Thoughts?

File: test/end-to-end/core.sh [hunk 9]
```diff
@@ -18,55 +19,55 @@ TEST_ASSETS="${TEST_ASSETS:-false}"
 function wait_for_app() {
   echo "[INFO] Waiting for app in namespace $1"
   echo "[INFO] Waiting for database pod to start"
-  wait_for_command "oc get -n $1 pods -l name=database | grep -i Running" $((60*TIME_SEC))
-  oc logs dc/database -n $1 --follow
+  os::cmd::try_until_text "oc get -n $1 pods -l name=database" 'Running'
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: This line fails for me locally even on master branch... between this issue and bash evaluating my comments as commands this wins for most annoying test.

**Comment by stevekuznetsov** [comment]
@liggitt Got green runs on a `devenv` machine.

[test]

**Comment by stevekuznetsov** [comment]
`TestAuthorizationSubjectAccessReview` flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8124/)

#6065

Re[test]

**Comment by stevekuznetsov** [comment]
flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8125/):
```
ERROR: gave up waiting for https://127.0.0.1:28443/healthz
```

#6031

**Comment by stevekuznetsov** [comment]
re[test]

**Comment by stevekuznetsov** [comment]
@liggitt Jenkins succeeded

**Comment by liggitt** [comment]
It's a Christmas miracle

File: test/end-to-end/core.sh [hunk 10]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
 
 # Process template and create
 echo "[INFO] Submitting application template json for processing..."
 STI_CONFIG_FILE="${ARTIFACT_DIR}/stiAppConfig.json"
 DOCKER_CONFIG_FILE="${ARTIFACT_DIR}/dockerAppConfig.json"
 CUSTOM_CONFIG_FILE="${ARTIFACT_DIR}/customAppConfig.json"
-oc process -n test -f examples/sample-app/application-template-stibuild.json > "${STI_CONFIG_FILE}"
-oc process -n docker -f examples/sample-app/application-template-dockerbuild.json > "${DOCKER_CONFIG_FILE}"
-oc process -n custom -f examples/sample-app/application-template-custombuild.json > "${CUSTOM_CONFIG_FILE}"
+os::cmd::expect_success "oc process -n test -f examples/sample-app/application-template-stibuild.json > ${STI_CONFIG_FILE}"
```
Comment (liggitt) [inline test/end-to-end/core.sh]: single quote the redirection destinations, e.g. `'${STI_CONFIG_FILE}'`

File: test/end-to-end/core.sh [hunk 11]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
 echo "[INFO] Streaming the logs from a deployment twice..."
-oc create -f test/fixtures/failing-dc.yaml
-tryuntil oc get rc/failing-dc-1
-oc logs -f dc/failing-dc
-wait_for_command "oc get rc/failing-dc-1 --template={{.metadata.annotations}} | grep openshift.io/deployment.phase:Failed" $((60*TIME_SEC))
-oc logs dc/failing-dc | grep 'test pre hook executed'
-oc deploy failing-dc --latest
-oc logs --version=1 dc/failing-dc
+os::cmd::expect_success 'oc create -f test/fixtures/failing-dc.yaml'
+os::cmd::expect_success 'tryuntil oc get rc/failing-dc-1'
```
Comment (liggitt) [inline test/end-to-end/core.sh]: try_until_success here?

File: test/end-to-end/core.sh (hunk 8 above)
Comment (liggitt) [inline test/end-to-end/core.sh]: move the TODO comment to its own line or remove it

File: test/end-to-end/core.sh [hunk 12]
```diff
@@ -278,57 +280,57 @@ echo "[INFO] Validating routed app response..."
...
 
 
 # Pod node selection
 echo "[INFO] Validating pod.spec.nodeSelector rejections"
 # Create a project that enforces an impossible to satisfy nodeSelector, and two pods, one of which has an explicit node name
-openshift admin new-project node-selector --description="This is an example project to test node selection prevents deployment" --admin="e2e-user" --node-selector="impossible-label=true"
+os::cmd::expect_success "openshift admin new-project node-selector --description='This is an example project to test node selection prevents deployment' --admin='e2e-user' --node-selector='impossible-label=true'"
 NODE_NAME=`oc get node --no-headers | awk '{print $1}'`
-oc process -n node-selector -v NODE_NAME="${NODE_NAME}" -f test/fixtures/node-selector/pods.json | oc create -n node-selector -f -
+os::cmd::expect_success "oc process -n node-selector -v NODE_NAME=${NODE_NAME} -f test/fixtures/node-selector/pods.json | oc create -n node-selector -f -"
```
Comment (liggitt) [inline test/end-to-end/core.sh]: keep single quotes around vars, e.g. `NODE_NAME='${NODE_NAME}'`

File: test/end-to-end/core.sh (hunk 11 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Good catch!

File: test/end-to-end/core.sh [hunk 13]
```diff
@@ -278,57 +280,57 @@ echo "[INFO] Validating routed app response..."
...
 
-oc project ${CLUSTER_ADMIN_CONTEXT}
+os::cmd::expect_success "oc project ${CLUSTER_ADMIN_CONTEXT}"
 # record the storage after pruning
-oc exec -p ${registry_pod} du /registry > ${LOG_DIR}/prune-images.after.txt
+os::cmd::expect_success "oc exec -p ${registry_pod} du /registry > ${LOG_DIR}/prune-images.after.txt"
 
 # make sure there were changes to the registry's storage
-[ -n "$(diff ${LOG_DIR}/prune-images.before.txt ${LOG_DIR}/prune-images.after.txt)" ]
+os::cmd::expect_failure "diff ${LOG_DIR}/prune-images.before.txt ${LOG_DIR}/prune-images.after.txt"
```
Comment (liggitt) [inline test/end-to-end/core.sh]: expect_code of 1...

File: test/end-to-end/core.sh (hunk 8 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @liggitt will remove once we get a response from @deads2k

**Comment by stevekuznetsov** [comment]
@liggitt single-quoted all redirect filenames, addressed other comments.

File: test/end-to-end/core.sh [hunk 14]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
-os::build:wait_for_start "test"
+os::cmd::try_until_text 'oc get -n test builds' 'Running'
+ 
 # Ensure that the build pod doesn't allow exec
-[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
-os::build:wait_for_end "test"
+BUILD_ID=$(oc get -n test builds  --output-version=v1 --template="{{with index .items 0}}{{.metadata.name}}{{end}}")
+os::cmd::expect_failure_and_text "oc rsh ${BUILD_ID}-build" 'forbidden'
+os::cmd::try_until_text 'oc get -n test builds' 'Complete' $((10*TIME_MIN))
+os::cmd::expect_success "oc build-logs -n test ${BUILD_ID} > '${LOG_DIR}/testbuild.log'" # TODO(skuznets): previously we were agnostic about this failing, why?
```
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Re-commenting since line was changed, comment on previous diff lost.

@deads2k previously we were agnostic about the `oc build-logs` failing - why? In what case do we want to continue the test when getting logs fails?

Comment (deads2k) [inline test/end-to-end/core.sh]: > @deads2k previously we were agnostic about the oc build-logs failing - why? In what case do we want to continue the test when getting logs fails?

No idea.  Expecting success seems more reasonable.

**Comment by stevekuznetsov** [comment]
Flake on `TestAuthorizationRestrictedAccessForProjectAdmins` [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8148/console):
```
F0104 12:55:11.819870   19941 controller.go:87] Unable to perform initial service nodePort check: unable to persist the updated port allocations: servicenodeportallocation "" cannot be updated: another caller has already initialized the resource
```

#6065 
re[test]

**Comment by stevekuznetsov** [comment]
e2e flaked with the nicely wrapped outcome [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8151/console):
```
FAILURE after 19.121s: hack/../test/end-to-end/core.sh:165: executing 'oc delete pod cli-with-token-2' expecting success: the command returned the wrong error code
There was no output from the command.
Standard error from the command:
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.12.107:4001] twice [last error: Unexpected HTTP status code]) [0]
```

#6447 

re[test]

**Comment by stevekuznetsov** [comment]
another flake [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8160/console):
```
FAILURE after 12.281s: hack/../test/end-to-end/core.sh:163: executing 'oc run cli-with-token-2 --attach --env=POD_NAMESPACE=test --image=openshift/origin:3a17cc0 --restart=Never -- cli whoami --loglevel=4 > '/tmp/origin/e2e/logs/cli-with-token2.log' 2>&1' expecting success: the command returned the wrong error code
There was no output from the command.
There was no error output from the command.
```
#6447 
re[test]

**Comment by stevekuznetsov** [comment]
re[test]

**Comment by stevekuznetsov** [comment]
@liggitt got green

File: test/end-to-end/core.sh [hunk 15]
```diff
@@ -278,57 +280,57 @@ echo "[INFO] Validating routed app response..."
...
 
-oc project ${CLUSTER_ADMIN_CONTEXT}
+os::cmd::expect_success "oc project ${CLUSTER_ADMIN_CONTEXT}"
 # record the storage after pruning
-oc exec -p ${registry_pod} du /registry > ${LOG_DIR}/prune-images.after.txt
+os::cmd::expect_success "oc exec -p ${registry_pod} du /registry > '${LOG_DIR}/prune-images.after.txt'"
 
 # make sure there were changes to the registry's storage
-[ -n "$(diff ${LOG_DIR}/prune-images.before.txt ${LOG_DIR}/prune-images.after.txt)" ]
+os::cmd::expect_failure "diff ${LOG_DIR}/prune-images.before.txt ${LOG_DIR}/prune-images.after.txt"
```
Comment (liggitt) [inline test/end-to-end/core.sh]: expect_code 1

**Comment by liggitt** [comment]
nit, LGTM otherwise... @deads2k, this seem like an improvement to you?

**Comment by stevekuznetsov** [comment]
@liggitt addressed comment

File: test/end-to-end/core.sh [hunk 16]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
 # services can end up on any IP.  Make sure we get the IP we need for the docker registry
 DOCKER_REGISTRY=$(oc get --output-version=v1beta3 --template="{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}" service docker-registry)
 
-registry="$(dig @${API_HOST} "docker-registry.default.svc.cluster.local." +short A | head -n 1)"
-[[ -n "${registry}" && "${registry}:5000" == "${DOCKER_REGISTRY}" ]]
+os::cmd::expect_success_and_text "dig @${API_HOST} docker-registry.default.svc.cluster.local. +short A | head -n 1 | xargs printf '%s:5000'" "${DOCKER_REGISTRY}"
 
 echo "[INFO] Verifying the docker-registry is up at ${DOCKER_REGISTRY}"
-wait_for_url_timed "http://${DOCKER_REGISTRY}/" "[INFO] Docker registry says: " $((2*TIME_MIN))
+os::cmd::try_until_success "curl --max-time 2 -fs http://${DOCKER_REGISTRY}/healthz" $((2*TIME_MIN))
```
Comment (deads2k) [inline test/end-to-end/core.sh]: tldr; revert this line

Not a fan of this change. `wait_for_url` does a bunch of stuff regarding certs.  Even though this is just `http`, I'd rather have `wait_for_url` and update that method to use our waiting functions.  Also, making `wait_for_url` do what we want instead of having two methods seems more reasonable.

File: test/end-to-end/core.sh [hunk 17]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
+os::cmd::expect_success "oc process -n docker -f examples/sample-app/application-template-dockerbuild.json > '${DOCKER_CONFIG_FILE}'"
+os::cmd::expect_success "oc process -n custom -f examples/sample-app/application-template-custombuild.json > '${CUSTOM_CONFIG_FILE}'"
 
 echo "[INFO] Back to 'test' context with 'e2e-user' user"
-oc login -u e2e-user
-oc project test
-oc whoami
+os::cmd::expect_success 'oc login -u e2e-user'
+os::cmd::expect_success 'oc project test'
+os::cmd::expect_success 'oc whoami'
```
Comment (deads2k) [inline test/end-to-end/core.sh]: No.  This was to provide info in the log

File: test/end-to-end/core.sh [hunk 18]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
+os::cmd::expect_success "oc process -n test -f examples/sample-app/application-template-stibuild.json > '${STI_CONFIG_FILE}'"
+os::cmd::expect_success "oc process -n docker -f examples/sample-app/application-template-dockerbuild.json > '${DOCKER_CONFIG_FILE}'"
+os::cmd::expect_success "oc process -n custom -f examples/sample-app/application-template-custombuild.json > '${CUSTOM_CONFIG_FILE}'"
 
 echo "[INFO] Back to 'test' context with 'e2e-user' user"
-oc login -u e2e-user
-oc project test
-oc whoami
+os::cmd::expect_success 'oc login -u e2e-user'
+os::cmd::expect_success 'oc project test'
```
Comment (deads2k) [inline test/end-to-end/core.sh]: no, provides info in the log we want.

File: test/end-to-end/core.sh [hunk 19]
```diff
@@ -84,193 +85,194 @@ wait_for_registry
...
-oc process -n custom -f examples/sample-app/application-template-custombuild.json > "${CUSTOM_CONFIG_FILE}"
+os::cmd::expect_success "oc process -n test -f examples/sample-app/application-template-stibuild.json > '${STI_CONFIG_FILE}'"
+os::cmd::expect_success "oc process -n docker -f examples/sample-app/application-template-dockerbuild.json > '${DOCKER_CONFIG_FILE}'"
+os::cmd::expect_success "oc process -n custom -f examples/sample-app/application-template-custombuild.json > '${CUSTOM_CONFIG_FILE}'"
 
 echo "[INFO] Back to 'test' context with 'e2e-user' user"
-oc login -u e2e-user
-oc project test
-oc whoami
+os::cmd::expect_success 'oc login -u e2e-user'
```
Comment (deads2k) [inline test/end-to-end/core.sh]: no, provides info in the log

**Comment by deads2k** [comment]
> nit, LGTM otherwise... @deads2k, this seem like an improvement to you?

I like the way this fixes `wait_for_command`, `expect_failure`, and `expect_*_and_text`.

@liggitt   I'm not a fan of the `os::cmd::expect_success` bits.  I feel pretty strongly, but I'm willing to be convinced.  I'd like to actually be able to see the output that we'd expect a customer script to have in this particular test.  We've learned things from that approach when we inspect failures, since the first look is "alright, what was happening when this line failed".

**Comment by stevekuznetsov** [comment]
@deads2k The mechanism for what you ask exists. We just use `VERBOSE=true os::cmd::whatever`. It doesn't seem like a good idea to whitelist "things we don't expect to fail" from the list of things we keep track of in case they start failing.

File: test/end-to-end/core.sh (hunk 16 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @deads2k, @liggitt and some others had this sort of conversation on another pull and we agreed that making wrapper methods that delegate to `os::cmd` is probably a bad idea, as it obfuscates what is going on. Please point out to me where `wait_for_url` does certificate things, as I did not see that. The only behavior I saw in the function was to run `curl` with some arguments over and over again, which this refactor makes much more obvious by pulling the arguments into the line.

Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: @deads2k This is `wait_for_url_timed`, which, paradoxically, does not delegate to `wait_for_url`, and therefore does no certificate actions.

**Comment by liggitt** [comment]
> @liggitt I'm not a fan of the os::cmd::expect_success bits. I feel pretty strongly, but I'm willing to be convinced. I'd like to actually be able to see the output that we'd expect a customer script to have in this particular test. We've learned things from that approach when we inspect failures, since the first look is "alright, what was happening when this line failed".

That does somewhat like noise here. I would want to turn on echoing of commands so we could see the commands as they are run interleaved with the results.

File: test/end-to-end/core.sh (hunk 16 above)
Comment (deads2k) [inline test/end-to-end/core.sh]: > @deads2k This is wait_for_url_timed, which, paradoxically, does not delegate to wait_for_url, and therefore does no certificate actions.

Revert this change for this pull and fix the `wait_for_url` stuff in another pull.

**Comment by deads2k** [comment]
> All the expect_success wrapping does somewhat like noise here. I would want to turn on echoing of commands so we could see the commands as they are run interleaved with the results.

@stevekuznetsov I'm pretty sure that this is @liggitt agreeing with me and suggesting that you do `set -x` for the `expect_success` bits.

File: test/end-to-end/core.sh (hunk 16 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: Could you elaborate on what " fix the `wait_for_url` stuff" means?

Comment (deads2k) [inline test/end-to-end/core.sh]: > Could you elaborate on what " fix the wait_for_url stuff" means?

I think we should have one helper function.  That helper should cleanly handle http and https.  Given the difficulty in setup, I want to avoid embedding the `curl` calls directly here.

**Comment by stevekuznetsov** [comment]
@liggitt Three reasons I do not want to use `set -x`:
 - it's not consistent with how we are getting that sort of data otherwise (*i.e.* `os::cmd`)
 - it gives us a strict subset of the data we get from `os::cmd`
 - it doesn't improve as `os::cmd` does, so we don't get more data when we need it for free

@deads2k was swayed by this argument [here](https://github.com/openshift/origin/pull/6592#discussion_r49319208). What do you think?

**Comment by liggitt** [comment]
I don't feel strongly

**Comment by stevekuznetsov** [comment]
@deads2k reverted changes for `wait_for_url_timed`

**Comment by deads2k** [comment]
Use `VERBOSE=true` on  every `expect_success` line in the main script.

**Comment by stevekuznetsov** [comment]
@deads2k Why is this necessary? I can understand something like `oc login` but `oc create -f some/imagestreams.json`? Should we log specifically interesting things to a logfile like we are doing with the pod exec information?

**Comment by stevekuznetsov** [comment]
@deads2k remember that the correct method for what I *think* you're trying to achieve (verbosity for Jenkins logs) is to amend the envars in vagrant-openshift or the `make` targets. We could also `export VERBOSE=true` and `unset VERBOSE` in this script... but why should I *need* to see verbose output for everything from the end-to-end test when I'm running it locally?

**Comment by stevekuznetsov** [comment]
Set everything to verbose... let's see what the log looks like.

**Comment by stevekuznetsov** [comment]
@deads2k output: https://gist.github.com/stevekuznetsov/705c43b6f63c2647878b

**Comment by stevekuznetsov** [comment]
I think that success was ailased with the logging bug:
```
[INFO] Verifying the docker-registry is up at 172.30.168.164:5000
hack/../test/end-to-end/core.sh: line 92: FRONTEND_IP: unbound variable
```
re[test]

**Comment by stevekuznetsov** [comment]
re[test]

File: test/end-to-end/core.sh [hunk 20]
```diff
@@ -9,28 +9,30 @@ set -o pipefail
...
 
   echo "[INFO] Waiting for database service to start"
-  wait_for_command "oc get -n $1 services | grep database" $((20*TIME_SEC))
+  os::cmd::try_until_text "oc get -n $1 services" 'database'
   DB_IP=$(oc get -n $1 --output-version=v1beta3 --template="{{ .spec.portalIP }}" service database)
 
   echo "[INFO] Waiting for frontend pod to start"
-  wait_for_command "oc get -n $1 pods | grep frontend | grep -i Running" $((120*TIME_SEC))
-  oc logs dc/frontend -n $1 --follow
+  os::cmd::try_until_text "oc get -n $1 pods" 'frontend.+Running'
```
Comment (deads2k) [inline test/end-to-end/core.sh]: if this starts flaking, remember that it waited for 2 minutes before.

Comment (liggitt) [inline test/end-to-end/core.sh]: Why change the timeout as part of this? Let's keep it the same

Comment (deads2k) [inline test/end-to-end/core.sh]: > Why change the timeout as part of this? Let's keep it the same

Doesn't two minutes seem crazy long?

File: test/end-to-end/core.sh [hunk 21]
```diff
@@ -84,193 +86,194 @@ wait_for_registry
 # services can end up on any IP.  Make sure we get the IP we need for the docker registry
 DOCKER_REGISTRY=$(oc get --output-version=v1beta3 --template="{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}" service docker-registry)
 
-registry="$(dig @${API_HOST} "docker-registry.default.svc.cluster.local." +short A | head -n 1)"
-[[ -n "${registry}" && "${registry}:5000" == "${DOCKER_REGISTRY}" ]]
+os::cmd::expect_success_and_text "dig @${API_HOST} docker-registry.default.svc.cluster.local. +short A | head -n 1 | xargs printf '%s:5000'" "${DOCKER_REGISTRY}"
 
 echo "[INFO] Verifying the docker-registry is up at ${DOCKER_REGISTRY}"
-wait_for_url_timed "http://${DOCKER_REGISTRY}/" "[INFO] Docker registry says: " $((2*TIME_MIN))
+wait_for_url_timed "http://${FRONTEND_IP}:5432" "[INFO] Frontend says: " $((2*TIME_MIN))
```
Comment (deads2k) [inline test/end-to-end/core.sh]: This looks wrong.  Why are you looking at the frontend_IP?

File: test/end-to-end/core.sh [hunk 22]
```diff
@@ -84,193 +86,194 @@ wait_for_registry
...
+os::cmd::expect_success 'oc logs --version=1 dc/failing-dc'
 
 echo "[INFO] Applying STI application config"
-oc create -f "${STI_CONFIG_FILE}"
+os::cmd::expect_success "oc create -f ${STI_CONFIG_FILE}"
 
 # Wait for build which should have triggered automatically
 echo "[INFO] Starting build from ${STI_CONFIG_FILE} and streaming its logs..."
-#oc start-build -n test ruby-sample-build --follow
-os::build:wait_for_start "test"
```
Comment (deads2k) [inline test/end-to-end/core.sh]: These helpers have general utility.  I'd like to keep using them to prevent rot.

File: test/end-to-end/core.sh [hunk 23]
```diff
@@ -84,193 +86,194 @@ wait_for_registry
...
+os::cmd::try_until_text 'oc get -n test builds' 'Running'
+ 
 # Ensure that the build pod doesn't allow exec
-[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
-os::build:wait_for_end "test"
```
Comment (deads2k) [inline test/end-to-end/core.sh]: These helpers have general utility. I'd like to keep using them to prevent rot.

File: test/end-to-end/core.sh [hunk 24]
```diff
@@ -84,193 +86,194 @@ wait_for_registry
...
-#echo "[INFO] Invoking generic web hook to trigger new docker build using curl"
-#curl -k -X POST $API_SCHEME://$API_HOST:$API_PORT/osapi/v1beta3/namespaces/docker/buildconfigs/ruby-sample-build/webhooks/secret101/generic && sleep 3
-#os::build:wait_for_end "docker"
-#wait_for_app "docker"
-
-#echo "[INFO] Applying Custom application config"
-#oc create -n custom -f "${CUSTOM_CONFIG_FILE}"
-#echo "[INFO] Invoking generic web hook to trigger new custom build using curl"
-#curl -k -X POST $API_SCHEME://$API_HOST:$API_PORT/osapi/v1beta3/namespaces/custom/buildconfigs/ruby-sample-build/webhooks/secret101/generic && sleep 3
-#os::build:wait_for_end "custom"
```
Comment (deads2k) [inline test/end-to-end/core.sh]: keep this.

File: test/end-to-end/core.sh (hunk 22 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: The helper:
 * repeatedly hits `oc get builds` and looks for the build to have started
 * prints that the build has started

I don't really see how that has more utility than this.

I *can* see that having a framework for being able to continue doing something, while providing both success *and* failure conditions in order to fail early is nice, and it seems like something that I could implement in `exec-assert` or, God help me, `test::cmd`, but if that's the only price we way for being able to integrate `test-end-to-end` with jUnit, so be it.

File: test/end-to-end/core.sh (hunk 23 above)
Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: See above. This helper has marginally more utility than the transcription.

Comment (deads2k) [inline test/end-to-end/core.sh]: > See above. This helper has marginally more utility than the transcription.

Its very obvious to read and to use.  I'd much rather have scripts describing intent like this.  At this point, its the only thing blocking your pull right?  Restore to use this and fight for it another day.

Comment (stevekuznetsov) [inline test/end-to-end/core.sh]: And it has the effect of aliasing test names because the calls to `os::cmd` will be in the helper func. This change overall isn't useful if it can't be useful for jUnit

**Comment by deads2k** [comment]
closed in favor of a version that gets 95% of the benefit without eliminating useful functions: https://github.com/openshift/origin/pull/7416
